components/nexus-repository/src/main/java/org/sonatype/nexus/repository/search/SearchFacetImpl.java

```diff
@@ -72,6 +75,27 @@ public SearchFacetImpl(final SearchService searchService,
   {
     this.searchService = checkNotNull(searchService);
     this.componentMetadataProducers = checkNotNull(componentMetadataProducers);
+    this.rebuildIndexLock = new ReentrantReadWriteLock();
+  }
+
+  @Override
+  @Guarded(by = STARTED)
+  public void rebuildIndex() {
+    rebuildIndexLock.writeLock().lock();
```

Comment:
Role of this lock is only to prevent put/delete to operate while index is being rebuilt. The lock use is non-blocking on their side, while rebuildIndex (called from a task) might block.